[AB-866] Launchdarkly

rodionchernyshov opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * LaunchDarkly feature-flagging across client and server with safe init, runtime evaluation and overrides.
  * Pluggable frontend observability (registry + adapter) with queued metrics for websockets, messaging, errors and performance.
  * UI build now produces sourcemaps with optional upload.

* **Chores**
  * Expanded publish pipeline: RC/versioning, PyPI publish + propagation wait and post-publish trigger.
  * Added dev tooling and Playwright; new npm scripts for sourcemap upload and combined build.

* **Style**
  * IDE metadata added to .gitignore.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

rodionchernyshov committed:
Polish LaunchDarkly observability integration

- Polish launchDarklyAdapter: reuse constants from utils, use logger instead of console
- Fix critical issues from CodeRabbit review:
  - Add cleanup for global error handlers to prevent duplicates
  - Fix incorrect LaunchDarkly package versions (@launchdarkly/observability: ^0.4.0, @launchdarkly/session-replay: ^0.4.5)
  - Add type safety check in evaluate_flag_with_override
- Remove all [Metrics] debug logs from frontendMetrics
- Standardize log format: use [LaunchDarkly] prefix consistently
- Remove duplicate initialization log (SDK already logs it)
- Add comprehensive LaunchDarkly utilities and client implementations

rodionchernyshov committed:
fix: code improvements

rodionchernyshov committed:
fix: code improvements

madeindjs commented on src/ui/src/core/index.ts:
```diff
@@ -159,6 +160,50 @@ export function generateCore() {
+			}
+		} catch (error) {
+			logger.error("Unexpected error initializing LaunchDarkly", error);
+		}
```
Can we move it to a new function `loadLaunchDarkly`

madeindjs commented on .github/workflows/publish.yml:
```diff
@@ -101,11 +101,38 @@ jobs:
+                npm run --if-present ld:upload-sourcemaps || echo "⚠️  Sourcemap upload skipped (LAUNCHDARKLY_ACCESS_TOKEN not set or script failed)"
+              continue-on-error: true
+
+            - name: Publish to PyPI
```
We don't want the `LAUNCHDARKLY_ACCESS_TOKEN` to be published on Pipy

rodionchernyshov replied:
Removed

madeindjs commented on src/writer/launchdarkly_utils.py:
```diff
@@ -0,0 +1,132 @@
+    Returns:
+        SDK key string if found, None otherwise
+    """
+    sdk_key = os.getenv("LAUNCHDARKLY_SDK_KEY")
```
Do we already have this ENV in prod / test ?

rodionchernyshov replied:
yes

rodionchernyshov replied:
the code already supports it

madeindjs commented on src/writer/launchdarkly_utils.py:
```diff
@@ -0,0 +1,132 @@
+    Returns:
+        LaunchDarkly environment: "Development", "Production", or "Test"
+        Defaults to "Development" if not set or invalid
+    """
+    env_raw = os.getenv("LAUNCHDARKLY_ENVIRONMENT", LAUNCHDARKLY_ENV_DEFAULT)
+    if env_raw in LAUNCHDARKLY_ENVIRONMENTS:
+        return env_raw
+
+    env_lower = env_raw.lower()
+    if env_lower in ("development", "dev"):
+        return LAUNCHDARKLY_ENV_DEVELOPMENT
+    elif env_lower in ("production", "prod"):
+        return LAUNCHDARKLY_ENV_PRODUCTION
+    elif env_lower == "test":
+        return LAUNCHDARKLY_ENV_TEST
+    else:
+        return LAUNCHDARKLY_ENV_DEFAULT
```
We also need to define this in prod / test / dev

madeindjs commented on src/ui/src/core/index.ts:
I think it'll be better to have a composable function that wrap everything

```ts
export function useObservabilityMetric() {
  async function initialize() {/*initialize launch darkly and `useConfigJs` */}

  async function updateSocketDuration() {/* ... */}

  // other stuffs form src/ui/src/observability/frontendMetrics.ts

  return { initialize, updateSocketDuration };
}
```

rodionchernyshov replied:
fixed

madeindjs commented on src/ui/src/utils/launchDarklyUtils.ts:
```diff
@@ -0,0 +1,176 @@
+		envClientId.trim() !== ""
+	) {
+		return envClientId;
+	}
+
+	return undefined;
+}
```
As I suggested earlier, it needs to be in a composable function. It can help you to ensure that `useConfigJs` has been called before, which is not done here.

So you can land in a situation when `getLaunchDarklyClientId` before `/env/config.js` has been loaded.

madeindjs commented on src/ui/vite.config.ts:
```diff
@@ -11,7 +11,15 @@ export default defineConfig({
+		),
+		LAUNCHDARKLY_ENVIRONMENT: JSON.stringify(
+			process.env.LAUNCHDARKLY_ENVIRONMENT || "Development",
+		),
```
Is this only for local testing ?

rodionchernyshov committed:
Refactor LaunchDarkly initialization into composable

- Move LaunchDarkly initialization to loadLaunchDarkly function
- Create useObservabilityMetric composable to wrap LaunchDarkly initialization
- Ensure useConfigJs is called before getLaunchDarklyClientId to prevent race conditions
- Use single shared instance of observability metric across codebase
- Remove LAUNCHDARKLY_ACCESS_TOKEN from PyPI publish step
- Simplify sessionId handling and make dependencies explicit

rodionchernyshov committed:
fix: code improvements

bybash commented on .github/workflows/publish.yml:
```diff
@@ -101,11 +101,38 @@ jobs:
+              run: |
+                WRITER_FRAMEWORK_VERSION=${{ steps.set_version.outputs.version }} npm run ui:build
+              env:
+                WRITER_FRAMEWORK_VERSION: ${{ steps.set_version.outputs.version }}
```
Do you need this part? You directly used version on line 110.

bybash commented on src/ui/src/core/index.ts:
```diff
@@ -87,6 +89,22 @@ export function generateCore() {
 
 	const activePageId = ref<Component["id"] | undefined>();
 
+	let observabilityMetricInstance: ReturnType<
+		typeof useObservabilityMetric
+	> | null = null;
```
Do you think that should be Ref object?

bybash commented on src/ui/src/core/index.ts:
```diff
@@ -337,6 +394,12 @@ export function generateCore() {
 				logger.error("Error parsing WebSocket message:", error);
+				const { incrementMetric, METRIC_NAMES } = await import(
+					"@/observability/frontendMetrics"
+				);
```
This seems repeating I think it is better to put it in a function and pass parameters of `incrementMetric` to that function, or import this at the top and don't use dynamic import. 
If you don't import dynamic, you don't need to make callbacks async.

bybash commented on src/ui/src/core/launchDarklyClient.ts:
```diff
@@ -0,0 +1,323 @@
+	context: LaunchDarklyContext,
+	clientSideId?: string,
+): Promise<void> {
+	if (initializationError) {
+		logger.warn(
+			"LaunchDarkly skipping initialization - previous error",
+			initializationError,
+		);
+		return;
+	}
```
Is this something what LD suggests? No way to recover it on next call?

bybash commented on src/ui/src/core/launchDarklyClient.ts:
```diff
@@ -0,0 +1,323 @@
+			.waitForInitialization(LAUNCH_DARKLY_INIT_TIMEOUT)
+			.catch(() => {
+				throw new Error("LaunchDarkly initialization timeout");
+			});
```
await and catch usage is a bit mix, you could use try catch maybe.

bybash commented on src/ui/src/core/launchDarklyClient.ts:
```diff
@@ -0,0 +1,323 @@
+}
+
+if (typeof window !== "undefined") {
+	(window as { __LD_CLIENT_MODULE__?: unknown }).__LD_CLIENT_MODULE__ = {
+		isLDInitialized: () => isInitialized && ldClient !== null,
+		getFlagValue: (flagKey: string, defaultValue: boolean) => {
+			if (!isInitialized || !ldClient) {
+				return defaultValue;
+			}
+			try {
+				return ldClient.variation(flagKey, defaultValue);
+			} catch (error) {
+				logger.warn(
+					`LaunchDarkly flag evaluation failed for ${flagKey}`,
+					error,
+				);
+				return defaultValue;
+			}
+		},
+	};
```
Someone from browser can override these functions and return true for both. I think it is better to use Object.freeze here.

bybash commented on src/ui/src/core/templateMap.ts:
```diff
@@ -27,6 +29,18 @@ export function setActiveFeatureFlags(flags: string[]) {
+				false,
+				false,
+			);
+		});
```
```suggestion
		return required.some((flag) => evaluateFlagWithOverride(
				flag,
				() => getFlagValue(flag, false),
				false,
				false,
			));
```

bybash commented on src/ui/src/utils/launchDarklyUtils.ts:
```diff
@@ -0,0 +1,150 @@
+
+export const LAUNCHDARKLY_ENV_DEFAULT: LaunchDarklyEnvironment =
+	LAUNCHDARKLY_ENV_DEVELOPMENT;
+
```
Do we need all of these? Can't we use the simple union type you have at the top?

bybash commented on src/ui/src/utils/launchDarklyUtils.ts:
```diff
@@ -0,0 +1,150 @@
+			if (
+				env === LAUNCHDARKLY_ENV_DEVELOPMENT ||
+				env === LAUNCHDARKLY_ENV_PRODUCTION ||
+				env === LAUNCHDARKLY_ENV_TEST
```
LAUNCHDARKLY_ENVIRONMENTS.includes(env)

rodionchernyshov committed:
fix: code improvements

rodionchernyshov committed:
fix: code improvements

rodionchernyshov committed:
fix: code improvements

rodionchernyshov committed:
fix: codereview improvements

madeindjs commented on src/ui/src/composables/useObservabilityMetric.ts:
```diff
@@ -0,0 +1,111 @@
+	}
+
+	function updateSocketDuration(connectStartTime: number): void {
+		if (!wf.isWriterCloudApp.value) {
+			return;
+		}
```
```suggestion
		if (!isLaunchDarklyInitialized) {
			return;
		}
```

madeindjs commented on src/ui/src/composables/useObservabilityMetric.ts:
```diff
@@ -0,0 +1,111 @@
+	return {
+		initialize,
+		updateSocketDuration,
+		isLaunchDarklyInitialized: () => isLaunchDarklyInitialized,
```
I guess that's a sign that `isLaunchDarklyInitialized` should be a `Ref<boolean>`. Then it becomes

```suggestion
		isLaunchDarklyInitialized: readonly(isLaunchDarklyInitialized),
```

madeindjs commented on src/ui/src/utils/launchDarklyUtils.ts:
```diff
@@ -0,0 +1,96 @@
+export function getLaunchDarklyClientId(): string | undefined {
+	return typeof window !== "undefined"
+		? window.__WRITER_APP_CONFIG__?.LAUNCH_DARKLY
+		: undefined;
```
`window.__WRITER_APP_CONFIG__` is a base64 string on prod (https://app.writer.com/env/config.js).


```suggestion
  if (typeof window !== "undefined") return;
  if (typeof window.__WRITER_APP_CONFIG__ === 'string') {
    try {
      return JSON.parse(atob(window.__WRITER_APP_CONFIG__)).LAUNCH_DARKLY
    } catch {
      return undefined
    }
  } elseif (typeof window.__WRITER_APP_CONFIG__ === 'object' && window.__WRITER_APP_CONFIG__) {
    return window.__WRITER_APP_CONFIG__.LAUNCH_DARKLY
  }
  return undefined;
```

madeindjs commented on src/ui/src/ambientTypes.ts:
```diff
@@ -7,4 +7,10 @@ declare module "vue" {
 
+declare global {
+	interface Window {
+		__WRITER_APP_CONFIG__?: WriterAppConfig;
```
`window.__WRITER_APP_CONFIG__` is a base64 string on prod (https://app.writer.com/env/config.js).

```suggestion
		__WRITER_APP_CONFIG__?: WriterAppConfig | string;
```

rodionchernyshov committed:
fix: codereview improvements

rodionchernyshov committed:
fix: codereview improvements

rodionchernyshov committed:
fix: codereview improvements

rodionchernyshov committed:
fix: codereview improvements

rodionchernyshov committed:
fix: codereview improvements

rodionchernyshov committed:
fix: codereview improvements

rodionchernyshov committed:
fix: codereview improvements

rodionchernyshov committed:
Merge branch 'dev' into AB-866

# Conflicts:
#	.github/workflows/publish.yml
